crates/indexer/src/indexer/fetcher.rs

```diff
@@ -0,0 +1,417 @@
+    context: Arc<Context>,
+    starting_version: u64,
+    retry_time_millis: u64,
+) -> (
+    aptos_types::transaction::Version,
+    aptos_types::transaction::Version,
+    aptos_types::account_config::NewBlockEvent,
+) {
+    loop {
+        match context.db.get_block_info_by_version(starting_version) {
+            Ok(info) => return info,
+            Err(err) => {
+                info!(
+                    "Failed to get block info by version {}, error: {:?}, retrying in {} ms",
+                    starting_version, err, retry_time_millis
+                );
+                tokio::time::sleep(Duration::from_millis(retry_time_millis)).await;
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl TransactionFetcherTrait for TransactionFetcher {
+    /// Fetches the next batch based on its internal version counter
+    async fn fetch_next_batch(&mut self) -> Vec<Transaction> {
+        self.transaction_receiver
+            .next()
+            .await
+            .expect("No transactions, producer of batches died")
+    }
+
+    fn fetch_ledger_info(&mut self) -> LedgerInfo {
+        self.context
+            .get_latest_ledger_info_wrapped()
+            .unwrap_or_else(|err| panic!("Failed to get ledger info: {}", err))
+    }
+
+    async fn set_version(&mut self, version: u64) {
+        if self.fetcher_handle.is_some() {
+            panic!("TransactionFetcher already started!");
+        }
+        self.starting_version = version;
+    }
+
+    async fn start(&mut self) {
+        if self.fetcher_handle.is_some() {
+            panic!("TransactionFetcher already started!");
+        }
+        let context = self.context.clone();
+        let transactions_sender = self.transactions_sender.take().unwrap();
+        let starting_version = self.starting_version;
+
+        let options2 = self.options.clone();
+        let fetcher_handle = tokio::spawn(async move {
+            let mut fetcher =
+                Fetcher::new(context, starting_version, options2, transactions_sender);
+            fetcher.run().await;
+        });
+        self.fetcher_handle = Some(fetcher_handle);
+    }
+}
+
+pub fn string_null_byte_replacement(value: &mut str) -> String {
+    value.replace('\u{0000}', "").replace("\\u0000", "")
+}
+
+pub fn recurse_remove_null_bytes_from_json(sub_json: &mut Value) {
+    match sub_json {
+        Value::Array(array) => {
+            for item in array {
+                recurse_remove_null_bytes_from_json(item);
+            }
+        }
+        Value::Object(object) => {
+            for (_key, value) in object {
+                recurse_remove_null_bytes_from_json(value);
+            }
+        }
+        Value::String(str) => {
+            if !str.is_empty() {
+                let replacement = string_null_byte_replacement(str);
+                *str = replacement;
+            }
+        }
+        _ => {}
+    }
+}
+
+pub fn remove_null_bytes_from_txn(txn: Transaction) -> Transaction {
```

Comment:
why is it that we need to do this here but not in firehose stream?

Reply:
we need to do it whenever we write to the DB